nipype/interfaces/base/tests/test_specs.py

```diff
@@ -63,6 +81,36 @@ def test_TraitedSpec_dynamic():
         assign_a_again
 
 
+def test_DynamicTraitedSpec_tab_completion():
+    def extract_func(list_out):
+        return list_out[0]
+
+    # Define interface
+    func_interface = Function(input_names=["list_out"],
+                             output_names=["out_file","another_file"],
+                             function=extract_func)
+    # Define node
+    list_extract = Node(Function(
+        input_names=["list_out"],output_names=["out_file"],
+        function=extract_func), name="list_extract")
+
+    # Check __all__ for interface inputs
+    expected_input = sorted(list_extract.inputs.editable_traits())
+    assert(sorted(func_interface.inputs.__all__) == expected_input)
+
+    # Check __all__ for node inputs
+    assert(sorted(list_extract.inputs.__all__) == expected_input)
+
+    # Check __all__ for node outputs
+    expected_output = sorted(list_extract.outputs.editable_traits())
+    assert(sorted(list_extract.outputs.__all__) == expected_output)
+
+    # Add trait and retest
+    list_extract.outputs.add_trait("added_out_trait","val")
```

Comment:
The problem is that the output spec is actually instantiated in `IOBase._outputs()`, which calls `Function._add_output_traits`, which only will add things in `self._output_names`. So to do what you want here, you should do:

`list_extract._interface._output_names.append('added_out_trait')`